components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/IdentityKeyStoreResolver.java

```diff
@@ -0,0 +1,374 @@
+     * @return Public Certificate of Primary or tenant keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public Certificate getCertificate(String tenantDomain) throws IdentityKeyStoreResolverException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        if (publicCerts.containsKey(String.valueOf(tenantId))) {
+            return publicCerts.get(String.valueOf(tenantId));
+        }
+
+        KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        Certificate publicCert;
+        try {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                publicCert = keyStoreManager.getDefaultPrimaryCertificate();
+            } else {
+                String tenantKeyStoreName = buildTenantKeyStoreName(tenantDomain);
+                publicCert = keyStoreManager.getCertificate(tenantKeyStoreName, tenantDomain);
+            }
+        } catch (Exception e) {
+            throw new IdentityKeyStoreResolverException("Error occurred when retrieving public certificate", e);
+        }
+
+        publicCerts.put(String.valueOf(tenantId), publicCert);
+        return publicCert;
+    }
+
+    /**
+     * Return Public Certificate of the Primary, tenant or custom keystore.
+     *
+     * @param tenantDomain      Tenant domain
+     * @param inboundProtocol   Inbound authentication protocol of the application
+     * @return Public Certificate of the Primary, tenant or custom keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public Certificate getCertificate(String tenantDomain, InboundProtocol inboundProtocol)
+            throws IdentityKeyStoreResolverException {
+
+        if (keyStoreMappings.containsKey(inboundProtocol)) {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain) ||
+                    keyStoreMappings.get(inboundProtocol).getUseInAllTenants()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Custom keystore configuration for " + inboundProtocol + " protocol.");
+                }
+
+                if (publicCerts.containsKey(inboundProtocol.toString())) {
+                    return publicCerts.get(inboundProtocol.toString());
+                }
+
+                String keyStoreName = buildCustomKeyStoreName(keyStoreMappings.get(inboundProtocol).getKeyStoreName());
+
+                try {
+                    int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+                    KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+                    Certificate publicCert = keyStoreManager.getCertificate(keyStoreName, null);
+                    publicCerts.put(inboundProtocol.toString(), publicCert);
+                    return publicCert;
+                } catch (Exception e) {
```

Comment:
Catch specific exceptions not generic ones "Exception"

Reply:
Same as https://github.com/wso2/carbon-identity-framework/pull/5794#discussion_r1687747359